modules/lvgl/lvgl.c

```diff
@@ -244,9 +244,12 @@ static int lvgl_init(void)
 	return 0;
 }
 
-BUILD_ASSERT(CONFIG_APPLICATION_INIT_PRIORITY < CONFIG_LV_Z_INPUT_INIT_PRIORITY);
+/* Display device needs to be ready */
+BUILD_ASSERT(CONFIG_KERNEL_INIT_PRIORITY_DEVICE < CONFIG_LV_Z_SYS_INIT_PRIORITY);
+/* Heap needs to be initialized first */
+BUILD_ASSERT(CONFIG_LV_Z_SYS_INIT_PRIORITY < CONFIG_LV_Z_INPUT_INIT_PRIORITY);
 #ifdef CONFIG_INPUT
 BUILD_ASSERT(CONFIG_INPUT_INIT_PRIORITY < CONFIG_LV_Z_INPUT_INIT_PRIORITY);
 #endif /* CONFIG_INPUT */
 
-SYS_INIT(lvgl_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+SYS_INIT(lvgl_init, POST_KERNEL, CONFIG_LV_Z_SYS_INIT_PRIORITY);
```

Comment:
if this is needed by "pseudo-drivers", then we have a problem with this design. Driver's dependencies on SYS_INIT should be avoided, and this is why I've always been very skeptical about using device-tree for non-hardware related stuff.

Reply:
Hmm I get your concern regarding this, putting pseudo devices into the device tree is unusual. 
The main benefits that imo it brings to the table are:
- it reduces boilerplate
- it allows for easy per instance configuration
- properties can be expressed in yaml files

A solution I could imagine is still using the yaml files for configuring them but move the instantiation to the source using macrobatics:
```
PSEUDO_DEVICE_DEFINE(compat, <per device specific config struct>);
```
The initialization would happen with a `SYS_INIT`. This would be my rough idea, needs to be validated ofc. What do you think?

Reply:
So the main problem I see in the current design is that LVGL init code (subsystem/module code) needs to be run before LVGL pseudo-devices, but after Kernel display devices. I'd consider a more reasonable design:

1. Kernel devices are initialized (no dependencies on 3rd party stuff)
2. LVGL is initialized, together with all its pseudo-devices.

This means we can leave Kernel alone and break its interdependency on a 3rd party module. I see a couple of solutions to this problem:

1. Keep using devicetree, but avoid using device model infrastructure. Overall, it looks like LVGL pseudo-devices can be freely initialized after all Kernel devices. So having an array of pseudo-device init calls run at lvgl_init would be enough (again, not using device model).
2. Avoid usage of devicetree and device model altogether. Simply provide a set of macros to statically define N pseudo-devices in C code. By using stuff like iterable sections, you can put all of them together and initialize them in lvgl_init.

(1) is a kind of "easy fix", (2) is more a change of paradigm, even though is what one usually sees in many other projects.

Reply:
How about: let the input device initialize but don't register them (i.e. drop `lv_indev_drv_register` from the individual drivers). Then in `lvgl_init`, after lvgl is initialized, cycle through all the various `zephyr,lvgl...input` devices and register them all.

You keep the tree, you can init all the input devices at `INPUT_INIT_PRIORITY` and drop a custom priority symbol, LVGL can initialze whenever after the devices.

Reply:
@fabiobaltieri I really like your solution, solves the problem at hand and actually simplifies priority handling. @gmarull would this be also a reasonable approach for you?

Reply:
makes sense